firebase-firestore/src/test/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProviderTest.java

```diff
@@ -0,0 +1,354 @@
+    // 3. The uncompleted task is completed.
+    // 4. The continuation added to the task is invoked and detects that tokenCounter changed, and
+    //       calls getToken() again.
+    // 5. The recursively-called getToken() calls calls internalAuthProvider.getAccessToken(), which
+    //       returns a completed task with the token "TestToken2".
+    // 6. The task returned from the original getToken() invocation completes.
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(getAccessTokenTaskCompletionSource.getTask());
+    Task<String> task = firebaseAuthCredentialsProvider.getToken();
+    verify(mockInternalAuthProvider).getAccessToken(anyBoolean());
+    idTokenListenerCaptor.getValue().onIdTokenChanged(mockInternalTokenResult);
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(Tasks.forResult(mockGetTokenResult2));
+    getAccessTokenTaskCompletionSource.setResult(mockGetTokenResult);
+
+    assertThat(task.isComplete()).isTrue();
+    assertThat(task.isSuccessful()).isTrue();
+    assertThat(task.getResult()).isEqualTo("TestToken2");
+  }
+
+  @Test
+  public void invalidateTokenShouldCauseGetTokenToForceRefresh() {
+    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
+        new FirebaseAuthCredentialsProvider(new ImmediateDeferred(mockInternalAuthProvider));
+    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(Tasks.forResult(mockGetTokenResult));
+
+    firebaseAuthCredentialsProvider.invalidateToken();
+    firebaseAuthCredentialsProvider.getToken();
+
+    verify(mockInternalAuthProvider).getAccessToken(true);
+  }
+
+  @Test
+  public void invalidateTokenShouldOnlyForceRefreshOnTheImmediatelyFollowingGetTokenInvocation() {
+    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
+        new FirebaseAuthCredentialsProvider(new ImmediateDeferred(mockInternalAuthProvider));
+    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(Tasks.forResult(mockGetTokenResult));
+
+    firebaseAuthCredentialsProvider.invalidateToken();
+    firebaseAuthCredentialsProvider.getToken();
+    firebaseAuthCredentialsProvider.getToken();
+
+    verify(mockInternalAuthProvider).getAccessToken(true);
+    verify(mockInternalAuthProvider).getAccessToken(false);
+  }
+
+  /** An implementation of {@link Deferred} whose provider never becomes available. */
+  private static final class UnavailableDeferred implements Deferred<InternalAuthProvider> {
+
+    @Override
+    public void whenAvailable(@NonNull DeferredHandler<InternalAuthProvider> handler) {}
+  }
+
+  /** An implementation of {@link Deferred} whose provider is always available. */
+  private static final class ImmediateDeferred implements Deferred<InternalAuthProvider> {
```

Comment:
Would it make sense to move these classes to separate package or utility class, so that they can be used from both test files? We could have something similar to the `Tasks` class with "Deferred.forX()" style factory methods.

I also wonder if `ImmediateDeferred` should be part of firebase-common.